src/reportSources/legacyFflogs/eventAdapter/prepull/action.ts

```diff
@@ -0,0 +1,34 @@
+import {Event, Events} from 'event'
+import {FflogsEvent} from 'fflogs'
+import {AdapterStep} from '../base'
+
+export class PrepullActionAdapterStep extends AdapterStep {
+	private finished = false
+
+	adapt(baseEvent: FflogsEvent, adaptedEvents: Event[]): Event[] {
+		const synthesizedEvents: Event[] = []
+
+		if (this.finished) {
+			return adaptedEvents
+		}
+
+		for (const event of adaptedEvents) {
```

Comment:
hm, this is still effectively iterating over the fflogs events one at a time, linearly. while it's _technically_ unlikely that there's >1 snapshot before the first action, if we assume a case that there _was_, we'd get
```
action
snapshot
action
snapshot
```

i... _might_ be over-thinking this, _but_ i think we should ensure the precast synths are _always_ before the first translated event.

spitball on how to do that;

```ts
class Etc {
	initialEvents?: Event[]
	adapt(baseEvent, adaptedEvents) {
		if (finished) { return ... }
		// grab a reference to the very first adapted events array
		if (initialEvents == null) { initialEvents = adaptedEvents }

		// logic
		
		// mutate the initial events array with the synthed events at the start
		initialEvents.splice(0, 0, ...synthedEvents)
	}
}
```

for that to work you'll need to tweak the adapter to something like

```ts
return events
	.map(baseEvent => theReduceStuffWeAlreadyHave)
	.flat()
```

(i.e. map -> flat, rather than flatmap)